File: Algo/Leetcode/085MaxRectangle.cpp
// O(m) space, O(mn) time
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int row = matrix.size();
        if(row == 0) return 0;
        int col = matrix[0].size();
        if(col==0) return 0;
        vector<int> height(col, 0);

        int max_area = 0;
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(matrix[i][j] == '1')
                    height[j] += 1;
                else
                    height[j] = 0;
            }
            max_area = max(max_area, maxRec(height));
        }
        return max_area;
    }

    int maxRec(vector<int>& height){
        stack<int> s;
        int n = height.size();
        vector<int> l(n);
        vector<int> r(n, n);

        for(int i=0; i<n; i++){
            while(!s.empty() && height[s.top()] > height[i]){
                r[s.top()] = i;
                s.pop();
            }
            l[i] = s.empty()?-1:s.top();
            s.push(i);
        }

        int max_area = 0;
        for(int i=0; i<n; i++){
            max_area = max(max_area, (r[i]-l[i]-1)*height[i]);
        }
        return max_area;
    }
};

File: Algo/Leetcode/085MaxRectangle.py
class Solution(object):
    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        n_row = len(matrix)
        n_col = len(matrix[0])
        mat = []
        for i in range(n_row):
            r = []
            for j in range(n_col):
                r.append(int(matrix[i][j]))
            mat.append(r)

        heights = mat[0]

        res = self.largestRectangleArea(heights)
        for i in range(1, n_row):
            new_height = mat[i]
            heights = [heights[j] + new_height[j] if new_height[j] > 0 else 0 for j in range(n_col)]
            res = max(res, self.largestRectangleArea(heights))

        return res

    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        st = []
        n = len(heights)
        l = [-1] * n
        r = [n] * n

        for i, h in enumerate(heights):
            while len(st) > 0 and heights[st[-1]] > h:
                r[st[-1]] = i  # keep lower right side as cur
                st.pop()
            if len(st) > 0:
                l[i] = st[-1]  # keep lower left side as next top
            st.append(i)

        res = 0
        for i, h in enumerate(heights):
            res = max((r[i] - l[i] - 1) * h, res)

        return res